# 配套实验一
## 实验要求
在《西瓜书》的习题3.4中这样写道：选择两个UCI数据集，比较10折交叉验证法与留一法所估计出的对数回归的错误率。

我们这里通过sklearn获取【鸢尾花数据集】与【葡萄酒数据集】数据集进行具体实验，这两组数据集都存在于UCI。

## 实验步骤

（1）使用sklearn加载数据集

（2）使用linear_model模块创建逻辑回归模型，注意参数选择和随机变量控制

（3）创建交叉验证（cross_val_score）

（4）将数据集与模型带入交叉验证，并得出结果

（5）创建留一法（LeaveOneOut）

（6）将数据集与模型带入使用留一法，并得出结果

## 参考答案
### 数据加载与分割

In [2]:
from sklearn import datasets

iris = datasets.load_iris() # scikit-learn 自带的 iris 数据集

### 十折交叉验证

In [41]:
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression

clf_lr = LogisticRegression(solver="liblinear",random_state=0)

scores_clf_lr_cv_acc = cross_val_score(clf_lr,iris.data,iris.target,cv=10,scoring='accuracy')
print("Accuracy: %0.4f (+/- %0.4f)" % (scores_clf_lr_cv_acc.mean(), scores_clf_lr_cv_acc.std()))

Accuracy: 0.9533 (+/- 0.0600)


### LOO

In [42]:
from sklearn.model_selection import LeaveOneOut
from sklearn.metrics import accuracy_score
import numpy as np

loo = LeaveOneOut()
loo.get_n_splits(iris.data)
mean_accuracy_score_list = []
for train_index, test_index in loo.split(iris.data):
    clf_lr.fit(iris.data[train_index], iris.target[train_index])
    prediction = clf_lr.predict(iris.data[test_index])
    mean_accuracy_score_list.append(accuracy_score(iris.target[test_index], prediction))
print("Accuracy: %0.4f (+/- %0.4f)" % (np.average(mean_accuracy_score_list),np.std(mean_accuracy_score_list)))

Accuracy: 0.9533 (+/- 0.2109)


### 结果对比

从预测结果上来说，十折交叉验证与LOO（留一法）平均正确率一样，都是0.9533。
因此又额外对比了标准差，从结果上来看可以看出交叉验证的标准差更小。


### 更换数据集并对别结果

In [43]:
wine = datasets.load_wine() # scikit-learn 自带的 wine 数据集

scores_clf_lr_cv_acc = cross_val_score(clf_lr,wine.data,wine.target,cv=10,scoring='accuracy')
print("Accuracy: %0.4f (+/- %0.4f)" % (scores_clf_lr_cv_acc.mean(), scores_clf_lr_cv_acc.std()))

Accuracy: 0.9556 (+/- 0.0544)


In [44]:
loo = LeaveOneOut()
loo.get_n_splits(wine.data)
mean_accuracy_score_list = []
for train_index, test_index in loo.split(wine.data):
    clf_lr.fit(wine.data[train_index], wine.target[train_index])
    prediction = clf_lr.predict(wine.data[test_index])
    mean_accuracy_score_list.append(accuracy_score(wine.target[test_index], prediction))
print("Accuracy: %0.4f (+/- %0.4f)" % (np.average(mean_accuracy_score_list),np.std(mean_accuracy_score_list)))

Accuracy: 0.9551 (+/- 0.2072)


### 结果对比
交叉验证比留一法正确率与标准差都更好一些。